sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -354,6 +356,148 @@ public void visitPrimitiveTransform(TransformHierarchy.Node node) { }
       @Override
       public void visitValue(PValue value, TransformHierarchy.Node producer) { }
     }
+
+    /**
+     * A {@link PipelineVisitor} used in {@link Pipeline#visualizePipeline()}
+     * to visualize a {@code Pipeline}.
+     *
+     */
+    class DisplayVisitor extends Defaults {
+      private List<String> pCollections = new ArrayList<>();
+      private List<String> collectionProducer = new ArrayList<>();
+      private List<TransformEdge> transformEdges = new ArrayList<>();
+
+      private List<Integer> endPoints = new ArrayList<>();
+      private boolean[][] edgeVector;
+
+      @Override
+      public void visitPrimitiveTransform(TransformHierarchy.Node transformTreeNode) {
```

Comment:
sorry that not quite sure 'composites'/'flattened' here, can you have more tips? are you referring to 
```java
transformTreeNode.isCompositeNode();
transformTreeNode.getEnclosingNode();
```